File: src/main/java/hiiragi283/material/api/item/MaterialItem.java
package hiiragi283.material.api.item;

import hiiragi283.material.HiiragiCreativeTabs;
import hiiragi283.material.api.material.HiiragiMaterial;
import hiiragi283.material.api.registry.HiiragiEntry;
import hiiragi283.material.api.shape.HiiragiShape;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.client.renderer.color.ItemColors;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import net.minecraftforge.oredict.OreDictionary;
import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public abstract class MaterialItem extends HiiragiItem {

    private final HiiragiShape shape;

    public MaterialItem(HiiragiShape shape) {
        super(shape.name(), 32767);
        this.shape = shape;
        setCreativeTab(HiiragiCreativeTabs.MATERIAL_ITEM);
    }

    @NotNull
    public HiiragiShape getShape() {
        return shape;
    }

    protected static final Consumer<HiiragiEntry.ITEM> MODEL_CONSUMER = item -> HiiragiUtil.setModelSame(item.getObject());

    public static MaterialItem create(HiiragiShape shape) {
        return create(shape, MODEL_CONSUMER, HiiragiUtil.getEmptyBiConsumer());
    }

    public static MaterialItem create(HiiragiShape shape, Consumer<HiiragiEntry.ITEM> model) {
        return create(shape, model, HiiragiUtil.getEmptyBiConsumer());
    }

    public static MaterialItem create(HiiragiShape shape, BiConsumer<HiiragiEntry.ITEM, HiiragiMaterial> recipe) {
        return create(shape, MODEL_CONSUMER, recipe);
    }

    public static MaterialItem create(
            HiiragiShape shape,
            Consumer<HiiragiEntry.ITEM> model,
            BiConsumer<HiiragiEntry.ITEM, HiiragiMaterial> recipe
    ) {

        return new MaterialItem(shape) {

            @Override
            protected void getRecipe(HiiragiEntry.ITEM item, HiiragiMaterial material) {
                recipe.accept(item, material);
            }

            @Override
            protected void getModel(HiiragiEntry.ITEM item) {
                model.accept(item);
            }
        };

    }

    //    Client    //

    @NotNull
    @Override
    @SideOnly(Side.CLIENT)
    public String getItemStackDisplayName(@NotNull ItemStack stack) {
        return HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .map(shape::getTranslatedName)
                .orElse(super.getItemStackDisplayName(stack));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void getSubItems(@NotNull CreativeTabs tab, @NotNull NonNullList<ItemStack> items) {
        if (!isInCreativeTab(tab)) return;
        HiiragiMaterial.REGISTRY.getValues().stream()
                .filter(material -> material.isIndexValid() && material.isSolid() && shape.isValid(material))
                .map(this::asItemStack)
                .sorted(Comparator.comparing(ItemStack::getMetadata))
                .filter(stack -> stack.getMetadata() > 0)
                .forEach(items::add);
    }

    //    HiiragiEntry    //

    @Override
    public void registerOreDict() {
        HiiragiMaterial.REGISTRY.getValues().stream()
                .filter(material -> material.isSolid() && shape.isValid(material))
                .forEach(material -> shape.getOreDicts(material).forEach(oreDict -> OreDictionary.registerOre(oreDict, asItemStack(material))));
    }

    @Override
    public void registerRecipe() {
        HiiragiMaterial.REGISTRY.getValues().stream()
                .filter(material -> material.isSolid() && shape.isValid(material))
                .forEach(material -> getRecipe(this, material));
    }

    protected abstract void getRecipe(HiiragiEntry.ITEM item, HiiragiMaterial material);

    @Override
    @SideOnly(Side.CLIENT)
    public void registerItemColor(ItemColors itemColors) {
        itemColors.registerItemColorHandler((stack, tintIndex) -> HiiragiMaterial.REGISTRY_INDEX.getValue(stack.getMetadata())
                .map(material -> material.color)
                .orElse(-1), this);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerModel() {
        getModel(this);
    }

    protected abstract void getModel(HiiragiEntry.ITEM item);

}

File: src/main/kotlin/hiiragi283/material/api/item/MaterialItem.kt
package hiiragi283.material.api.item

import hiiragi283.material.api.ingredient.ItemIngredient
import hiiragi283.material.api.machine.MachineType
import hiiragi283.material.api.material.HiiragiMaterial
import hiiragi283.material.api.part.HiiragiPart
import hiiragi283.material.api.part.PartConvertible
import hiiragi283.material.api.shape.HiiragiShape
import hiiragi283.material.init.HiiragiCreativeTabs
import hiiragi283.material.init.HiiragiRegistries
import hiiragi283.material.init.HiiragiShapes
import hiiragi283.material.item.ItemShapePattern
import hiiragi283.material.recipe.MachineRecipe
import hiiragi283.material.util.hiiragiLocation
import hiiragi283.material.util.isSameWithoutCount
import hiiragi283.material.util.itemStack
import hiiragi283.material.util.setModelSame
import net.minecraft.client.renderer.color.IItemColor
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.minecraftforge.oredict.OreDictionary

open class MaterialItem(final override val shape: HiiragiShape) : HiiragiItem(
    shape.name,
    Short.MAX_VALUE.toInt()
), PartConvertible.ITEM {

    init {
        creativeTab = HiiragiCreativeTabs.MATERIAL_ITEM
    }

    //    Client    //

    @SideOnly(Side.CLIENT)
    override fun getItemStackDisplayName(stack: ItemStack): String =
        HiiragiRegistries.MATERIAL_INDEX.getValue(stack.metadata)
            ?.let(shape::getTranslatedName)
            ?: super.getItemStackDisplayName(stack)

    @SideOnly(Side.CLIENT)
    override fun getSubItems(tab: CreativeTabs, subItems: NonNullList<ItemStack>) {
        if (!isInCreativeTab(tab)) return
        HiiragiRegistries.MATERIAL_INDEX.getValues()
            .filter(HiiragiMaterial::isValidIndex)
            .filter(shape::isValid)
            .map(::itemStack)
            .forEach(subItems::add)
    }

    //    HiiragiEntry    //

    override fun onRegister() {
        HiiragiRegistries.MATERIAL_ITEM.register(shape, this)
    }

    override fun registerOreDict() {
        HiiragiRegistries.MATERIAL_INDEX.getValues()
            .filter(shape::isValid)
            .forEach { material: HiiragiMaterial ->
                shape.getOreDicts(material).forEach {
                    OreDictionary.registerOre(it, itemStack(material))
                }
            }
    }

    override fun registerRecipe() {
        HiiragiRegistries.MATERIAL_INDEX.getValues()
            .filter { material: HiiragiMaterial -> material.isSolid() && shape.isValid(material) }
            .forEach { addRecipes(it) }
    }

    open fun addRecipes(material: HiiragiMaterial) {}

    fun addGrinderRecipe(material: HiiragiMaterial) {
        if (shape.scale < 144) return
        val part: HiiragiPart = shape.getPart(material)
        MachineRecipe.buildAndRegister(MachineType.GRINDER, hiiragiLocation(part.toString())) {
            inputItems.add(ItemIngredient.Parts(part))
            outputItems.add(
                HiiragiRegistries.MATERIAL_ITEM
                    .getValue(HiiragiShapes.DUST)
                    ?.item()
                    ?.itemStack(part)
                    ?: ItemStack.EMPTY
            )
        }
    }

    fun addMetalFormerRecipe(material: HiiragiMaterial, inputCount: Int = 1, outputCount: Int = 1) {
        if (!HiiragiShapes.INGOT.isValid(material) || !shape.isValid(material)) return
        val part: HiiragiPart = shape.getPart(material)
        MachineRecipe.buildAndRegister(
            MachineType.METAL_FORMER,
            hiiragiLocation(part.toString())
        ) {
            inputItems.add(ItemIngredient.Parts(HiiragiShapes.INGOT, part.material, inputCount))
            inputItems.add(
                ItemIngredient.Custom(
                    stacks = { listOf(ItemShapePattern.getItemStack(part.shape)) },
                    predicate = { stack -> stack.isSameWithoutCount(ItemShapePattern.getItemStack(part.shape)) },
                    process = ItemIngredient.CATALYST_PROCESS
                )
            )
            outputItems.add(part.getItemStack(outputCount))
        }
    }

    @SideOnly(Side.CLIENT)
    override fun getItemColor(): IItemColor = HiiragiMaterial.ITEM_COLOR

    @SideOnly(Side.CLIENT)
    override fun registerModel() = this.setModelSame()

}